R/derive_param_bor.R

```diff
@@ -0,0 +1,518 @@
+    required_vars = quo_c(subject_keys)
+  )
+
+  filter_source <- assert_filter_cond(arg = enquo(filter_source))
+
+  assert_integer_scalar(
+    arg      = ref_start_window,
+    subset   = "non-negative",
+    optional = TRUE
+  )
+
+  assert_logical_scalar(arg = missing_as_ne)
+
+  assert_function(arg = aval_fun)
+
+  assert_vars(arg = subject_keys)
+
+  assert_varval_list(
+    arg               = set_values_to,
+    required_elements = c("PARAMCD")
+  )
+
+  assert_param_does_not_exist(
+    dataset = dataset,
+    param   = quo_get_expr(set_values_to$PARAMCD)
+  )
+
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  # filter_pd and filter_source: Filter source dataset using filter_source----
+  # argument and also filter data after progressive disease with filter_pd
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  if (!is.null(source_pd)) {
+    dataset_filter <- dataset %>%
+      filter_pd(
+        filter          = !!filter_source,
+        source_pd       = source_pd,
+        source_datasets = source_datasets,
+        subject_keys    = subject_keys
+      )
+  } else {
+
+    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    # filter_source: Filter using filter_source argument ----
+    # This would also be used to filter out records from dataset that are greater
+    # than e.g. ADSL.TRTSDT
+    # Not filtering data after progressive disease with filter_pd
+    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+    dataset_filter <- dataset %>%
+      filter(!!enquo(filter_source))
+  }
+
+  # Error if filter results in 0 records
+  if (nrow(dataset_filter) == 0) {
+    err_msg <- sprintf(
+      "dataframe passed into %s argument with the filter %s has 0 records",
+      "dataset",
+      deparse(quo_get_expr(filter_source))
+    )
+
+    abort(err_msg)
+  }
+
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  # Create Sort Order for Selection of Minimum Later ----
+  #
+  #   Note:
+  #   1. All `CR`, `PR` and `PD` response records are considered for Best Overall
+  #      Response.
+  #
+  #   2. All `SD` or `NON-CR/NON-PD` records where `ADT` >= `reference_date` +
+  #     `ref_start_window` are also considered for Best Overall Response.
+  #
+  #  3. Subjects with **ONLY** `SD` or `NON-CR/NON-PD` records where `ADT` <
+  #     `reference_date` + `ref_start_window` would give Best Overall Response
+  #     of `NE`.
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  dataset_ordered <- dataset_filter %>%
+    mutate(
+      tmp_order = case_when(
+        AVALC %in% c("CR") ~ 1,
+        AVALC %in% c("PR") ~ 2,
+        AVALC %in% c("SD") ~ 3,
+        AVALC %in% c("NON-CR/NON-PD") ~ 4,
+        AVALC %in% c("PD") ~ 5,
+        AVALC %in% c("NE") ~ 6,
+        is.null(AVALC) ~ 7
+      ),
+      AVALC = if_else(
```

Comment:
`AVALC` must be updated first. Consider for example
```
adsl <- tribble(
  ~USUBJID, ~TRTSDTC,
  "1",      "2020-01-01"
) %>%
  mutate(
    TRTSDT = ymd(TRTSDTC),
    STUDYID = "XX1234"
  )
ovr_obs <- tribble(
  ~USUBJID, ~ADTC,        ~AVALC,
  "1",      "2020-01-01", "SD",
  "1",      "2020-02-01", "NON-CR/NON-PD"
) %>%
  mutate(PARAMCD = "OVR")

```

Reply:
My bad, trying to get this done before I leave, thank you for spoting this, my tests still passed for some reason (need to update to ensure this does not hapen again)